src/control/system/ranklist.go

```diff
@@ -73,6 +73,24 @@ func NewRankSet(stringRanks string) (*RankSet, error) {
 	return &rs, nil
 }
 
+// Add adds rank to an existing RankSet.
+func (rs *RankSet) Add(rank Rank) error {
+	rs.HostSet.Lock()
+	defer rs.HostSet.Unlock()
```

Comment:
I'm not 100% sure about what's happening behind the scenes on this. You're taking a lock on the embedded `HostSet` which gets removed in the defer, but you're also creating a new `HostSet` to replace the old one. In this case it's probably fine; I assume that there's a reference to the old one held in the defer closure.

It might be clearer to just create a separate `sync.RWMutex` for `RankSet` though. I think that this could also possibly be a problem down the road if a new method is added to `RankSet` with code that also has to run under a lock -- there could be a race as to which lock is taken (one in the new `HostSet` or the old one).

Reply:
I agree with the idea of creating a new mutex for the RankSet. As it is, it should be safe, but it's a little confusing to reason about.